### ISSUE TITLE
problem with OTA update

### ISSUE THREAD

**Issue by alandolt** [stelgenhof/AiLight#43]
Hello

First thank you very much for the development of the AiLight. It works great on my bulb and thanks to it I could develop an alarm clock which is controlled over a web application (with the Rest API). Thanks. 

The upload over serial connection works without problem, but unfortanetely the OTA update doesn't work at all. 
When I'm trying to do an ota update with visual studio code ('pio run --target upload -e  prod-ota --upload-port 192.168.1.198'), I'm getting the following error while uploading:

`21:02:14 [DEBUG]: Options: {'timeout': 10, 'esp_ip': '192.168.1.198', 'host_port': 45376, 'image': '.pioenvs\\prod-ota\\firmware.bin', 'host_ip': '0.0.0.0', 'auth': 'hinotori --port=8266', 'esp_port': 8266, 'spiffs': False, 'debug': True, 'progress': True}
21:02:14 [INFO]: Starting on 0.0.0.0:45376
21:02:14 [INFO]: Upload size: 397936
Sending invitation to 192.168.1.198
Authenticating...FAIL
21:02:14 [ERROR]: Authentication Failed`

I didn't change the standard platformeio.ini and config.h so the port should be 8266 and the password should be hinotori as described in the error message. The webinterface of the bulb listen on 192.168.1.198, so the ip is correct. 

The following things I've tried:

- The upload over ota doesn't work for me in the 0.5 version and the actual 0.6 development version
- the firewall of windows wasn't on while testing
- In the platformeio.ini file I've tried to add an = after port in the upload flags, but it didn't work either
` upload_flags = --auth=${common.ota_password} --port=${common.ota_port}`

- I've tried the other ota build (dev-ota) too, but unfortanetely without result. 
- I've tried to add the --port in upload_flags on a new line after --auth=, but then the compilation didn't work. 

So now I don't know what I can try anymore, I'm very glad for any help. 
thanks you very much. 
Greetings 
Cat Turbo

PS: in the attachement you find my config.h and my platformeio.ini file, but they are more or less the same as the original ones
[platformeio and config.zip](https://github.com/stelgenhof/AiLight/files/2337579/platformeio.and.config.zip)

**Comment by brahmafear** [comment]
Hi CatTurbo,

I had same issue. Despite the error saying “authentication failure”, the reason the the failure was insufficient flash space on the bulb. Ie, flash was not big enough to hold 2 copies of code (required for  ota updates). You can see this error message by running some of the command line programs manually. 

You might try doing the typical, 2 stage OTA update — do an ota upload of some very stripped down code that does nothing but wait for an ota update, then do an ota update of the newer AiLight code. 

Hope this helps.

**Comment by alandolt** [comment]
Hello brahmafear
First thank you very much for your fast answer. I think you are completely right, my bulb has 1MB of flash, so space is IMHO here the problem. 
About the 2 stage update. I've tried it with the standard android OTA sketch for ESP8266

The upload worked. In the next step I've tried to upload another code (the ailight code), but the PIO gave me a memory failure error. In the Arduino IDE the flash also didn't work too. 
 
Additionally I've got a huge problem to connect the bulb with a serial interface, cause the last time I desoldered the calbes, I've accidentally damaged the RX solder pad on the bulb (now it is broken and it don't have a connection to the ESP chip anymore) . So this time I had to make a connection directly to the ESP chip with a multimeter probe . It worked, but I don't think it will work a second time ;)
So do you think that the OTA update should work normally with the arduino code posted above or do I have to use another code for a two stage flash?
Thank you very much for a response


[BasicOTA_test.zip](https://github.com/stelgenhof/AiLight/files/2343695/BasicOTA_test.zip)

**Comment by brahmafear** [comment]
Could you provide the exact error messages you received?

**Comment by alandolt** [comment]
Good evening brahmafear

first thanks you again for your response. In the meantime I think I've found the bug. 
I've first reflashed the Ai-Light firmware on the bulb (over a serial connection). Then I've flashed the Arduino OTA code on the bulb with platforme IO (over OTA update). There it seems to be important that there isn't an upload port specified in the platformeio.ini file. 
```upload_flags = --auth="here your password" ```
and not 
```upload_flags = --auth="here your password --port here_your_port```
as specified in the platformeio.ini file of the ai-light project. 

As soon as I specify a port in the platform.ini file, I'm getting an authentification error. So in the next step I've flashed the ai-light code on the bulb, also with no upload port specified. It worked. 
So just for fun I've tried after to flash directly a new version of the Ai-light firmware on the bulb when there is an existing ai-light code on the bulb. It works (no memory, no space error), but here the important thing too is to not specify a upload port. Then it works without authentification error and I can it flash too without the two steps flash. 

After some more testing I've discovered that the port should be defined on a new line with two blanks as intending and a = after port:
 
```
upload_flags = --auth=${common.ota_password} 
   --port=${common.ota_port}
```

This is also described in the docs, here
http://docs.platformio.org/en/latest/platforms/espressif8266.html
and here  https://docs.platformio.org/en/latest/projectconf/section_env_upload.html 


With this modification it works without problem.

**Comment by brahmafear** [comment]
Great diagnostics!

@stelgenhof -- Can we get that change put in to next release?  I'd submit a
PR, but I'm not sure what to base on -- last release was a year back.

**Comment by stelgenhof** [comment]
@CatTurbo @brahmafear Thanks for catching this issue. I can do a fix for you no problem. 

Yes, I know the last release was a year back :( Lame excuse, but busy with other projects at the same time, haha. 

There are some minor changes/fixes done since, so maybe I can do a minor release (v0.5.1) soon.

**Comment by stelgenhof** [comment]
Solved in the develop branch. You may need to adjust your own platformio.ini file yourself. 

I will do a 0.5.1 release probably end of this week. 

Cheers! Sacha

**Comment by stelgenhof** [comment]
Included in the just released version 0.6.0.
